nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/groovy/org/apache/nifi/properties/sensitive/aws/kms/AWSKMSSensitivePropertyProviderIT.groovy

```diff
@@ -0,0 +1,253 @@
+            propProvider = new AWSKMSSensitivePropertyProvider("")
+        }
+        assert msg =~ "The key cannot be empty"
+
+
+        assert propProvider == null
+
+        msg = shouldFail(com.amazonaws.services.kms.model.NotFoundException) {
+            propProvider = new AWSKMSSensitivePropertyProvider("bad key")
+            propProvider.protect("value")
+        }
+        assert msg =~ "Invalid keyId"
+    }
+
+    /**
+     * These tests show that the provider with known keys can round-trip protect + unprotect random, generated text.
+     */
+    @Test
+    void testShouldProtectAndUnprotectValues() throws Exception {
+        SensitivePropertyProvider propProvider
+        String plainText
+
+        knownGoodKeys.each { k ->
+            propProvider = new AWSKMSSensitivePropertyProvider(k)
+            assert propProvider != null
+
+            byte[] randBytes = new byte[1024]
+            new SecureRandom().nextBytes(randBytes)
+            plainText = randBytes.encodeBase64()
+
+            assert plainText != null
+            assert plainText != ""
+
+            assert plainText == propProvider.unprotect(propProvider.protect(plainText))
+        }
+    }
+
+    /**
+     * These tests show that the provider cannot encrypt empty values.
+     */
+    @Test
+    void testShouldHandleProtectEmptyValue() throws Exception {
+        SensitivePropertyProvider propProvider
+        final List<String> EMPTY_PLAINTEXTS = ["", "    ", null]
+
+        knownGoodKeys.each { k ->
+            propProvider = new AWSKMSSensitivePropertyProvider(k)
+            assert propProvider != null
+
+            EMPTY_PLAINTEXTS.each { String emptyPlaintext ->
+                def msg = shouldFail(IllegalArgumentException) {
+                    propProvider.protect(emptyPlaintext)
+                }
+                assert msg == "Cannot encrypt an empty value"
+            }
+        }
+    }
+
+    /**
+     * These tests show that the provider cannot decrypt invalid ciphertext.
+     */
+    @Test
+    void testShouldUnprotectValue() throws Exception {
+        SensitivePropertyProvider propProvider
+        final List<String> BAD_CIPHERTEXTS = ["any", "bad", "value"]
+
+        knownGoodKeys.each { k ->
+            propProvider = new AWSKMSSensitivePropertyProvider(k)
+            assert propProvider != null
+
+            // text that cannot be decoded values throw a bouncy castle exception:
+            BAD_CIPHERTEXTS.each { String emptyPlaintext ->
+                def msg = shouldFail(org.bouncycastle.util.encoders.DecoderException) {
+                    propProvider.unprotect(emptyPlaintext)
+                }
+                assert msg != null
+            }
+
+            // Empty string throws a different exception:
+            def msg = shouldFail(com.amazonaws.services.kms.model.AWSKMSException) {
+                propProvider.unprotect("")
+            }
+            assert msg != null
+        }
+    }
+
```

Comment:
I would add a test which attempts (and fails as expected) to decrypt valid Base64 content that is _not_ encrypted with the AWS key to exercise that the provider correctly identifies when invalid input (that is still Base64) fails to decrypt.

Reply:
Added "good base64 string with bad ciphertext" test.